Userland/Utilities/tail.cpp

```diff
@@ -20,27 +21,38 @@ static ErrorOr<void> tail_from_pos(Core::File& file, off_t startline)
     return {};
 }
 
-static ErrorOr<off_t> find_seek_pos(Core::File& file, int wanted_lines)
+static ErrorOr<off_t> find_seek_pos(Core::File& file, int wanted_lines, bool start_from_end)
 {
-    // Rather than reading the whole file, start at the end and work backwards,
-    // stopping when we've found the number of lines we want.
-    off_t pos = TRY(file.seek(0, SeekMode::FromEndPosition));
-
-    off_t end = pos;
     int lines = 0;
 
-    for (; pos >= 1; pos--) {
-        TRY(file.seek(pos - 1, SeekMode::SetPosition));
+    if (start_from_end) {
+        off_t pos = TRY(file.seek(0, SeekMode::FromEndPosition));
+        off_t end = pos;
+        for (; pos >= 1; pos--) {
+            TRY(file.seek(pos - 1, SeekMode::SetPosition));
+
+            auto ch = TRY(file.read_value<u8>());
+            if (ch == '\n' && (end - pos) > 0) {
+                lines++;
+                if (lines == wanted_lines)
+                    break;
+            }
+        }
+        return pos;
+    }
+
+    size_t file_size = TRY(file.size());
 
+    for (size_t i = 0; i < file_size; i++) {
         auto ch = TRY(file.read_value<u8>());
-        if (ch == '\n' && (end - pos) > 0) {
+        if (ch == '\n') {
             lines++;
             if (lines == wanted_lines)
                 break;
         }
     }
 
-    return pos;
+    return file.tell().value();
```

Comment:
Calling `value()` here stops any errors from the `tell()` call from being propagated correctly. You should just be able to return `file.tell()`.

Also, is the `tell()` call necessary here? Could you return `i` in this case? If it's possible to do this, I would rename `i` to `pos` to be consistent with the code above.

Reply:
> Calling value() here stops any errors from the tell() call from being propagated correctly. You should just be able to return file.tell().

Yes.

> Also, is the tell() call necessary here? Could you return i in this case? If it's possible to do this, I would rename i to pos to be consistent with the code above.

I could return `i + 1` instead (`read_value<u8>()` does advance the cursor).

> I think this will skip the first character if you do tail -n +0.

No, this has nothing to do with this code. This happens in the non-seekable code (and yes, it currently does not check for skip lines being 0)

Reply:
> No, this has nothing to do with this code. This happens in the non-seekable code (and yes, it currently does not check for skip lines being 0)

Ah yes, this obviously isn't the right place. Sorry for the mix-up!